### ISSUE TITLE
support for @inlineCallbacks

### ISSUE THREAD

**Issue by obeleh** [crossbario/txaio#62]
Would it be possible to support inlineCallbacks?
I would like to do something similar to the code below:

```python
from __future__ import print_function
import txaio
from asyncio.futures import Future
from asyncio.coroutines import coroutine
import asyncio
txaio.use_asyncio()
from twisted.internet.defer import Deferred
from twisted.internet.defer import inlineCallbacks
from twisted.internet import reactor


@inlineCallbacks
def test1():
    dfd = Deferred()
    reactor.callLater(1, dfd.callback, 'Test1 OK')
    print('Test1 started')
    res = yield dfd
    print(res)


@inlineCallbacks
def test2():
    fut = Future()
    reactor.callLater(1, fut.set_result, 'Test1 OK')
    print('Test2 started')
    res = yield fut
    print(res)

@coroutine
def test3():
    fut = Future()
    reactor.callLater(1, fut.set_result, 'Test1 OK')
    print('Test3 started')
    res = yield fut
    print(res)

test1()
test2()
test3()

asyncio.get_event_loop().run_forever()
```

results when I used the asyncio eventloop
```
Test1 started
Test2 started
<Future pending>
```

results when using `reactor.run()`
```
Test1 started
Test2 started
<Future pending>
Test1 OK
```

**Comment by obeleh** [comment]
I'm sorry an I doing things you're not supposed to do in this library? I was under the impression I could intermix asyncio with python code.

**Comment by oberstet** [comment]
No, you cannot mix both in a single program, as each has its own event loop. You can write a program that then can run unmodified under either asyncio or Twisted - but not at the same time.

**Comment by meejah** [comment]
Conceivable, it might be possible to add txaio support for some sort of decorator-wrapper over @inlineCallbacks/@coroutine but of course only one at a time. That is, instead of ``@inlineCallback``s or ``@coroutine`` you'd decorate with ``txaio.coroutine`` or similar.

However, we haven't needed this in Autobahn...

Also, I'm not sure how you'd get around the ``yield from`` difference: any "txaio" decorator would have to work in Python2.7 with Twisted, so there'd need to be some way to *always* write code with a plain ``yield`` but have it work properly with asyncio in Python3.x -- and I'm not sure if that's possible, so that'd be the place to start ;)

If you really want to try mixing twisted/asyncio I think https://github.com/itamarst/txtulip is the place to start.

**Comment by oberstet** [comment]
@meejah You can't get over the `yield` **syntax** differences between Python 2 and Python 3, which is the very reason we don't use it at all in Autobahn, but plain Deferreds/Futures. Because of this, there is little value IMO in papering over `@inlineCallbacks/@coroutine` with a `txaio.coroutine`.

> The new `async`/`await` in Python 3.5 is also completely new, breaking syntax. A program written using that syntax cannot run < 3.5. At least everything I just said is to the best of my knowledge, and I did dig into this quite a bit .. but who knows.

**Comment by meejah** [comment]
@oberstet yeah, I agree that it's Way Safer to just stick w/ callback-style things.

I was just thinking that there **might** be some way for txaio to make plain ``yield`` in Python3 behave like ``yield from`` as far as Futures are concerned -- but, I haven't explored that at all and it very well could be impossible. And then, yes, there's also the ``async`` syntax that obviously also can't be magically made to appear in Python2 ;)

The bottom line is really: txaio exists to make it *possible* to support py2/py3 and Twisted/asyncio/trollius from one codebase -- but imposes certain limitations (i.e. must do callback-style using Futures/Deferreds).

@obeleh are there documentation improvements we could make such that the above is more obvious...?

**Comment by obeleh** [comment]
It was this part that helped me make to wrong conclusions:

```
txaio is a helper library for writing code that runs unmodified on both Twisted and asyncio / Trollius.

This is like six, but for wrapping over differences between Twisted and asyncio so one can write code that runs unmodified on both (aka source code compatibility). In other words: your users can choose if they want asyncio or Twisted as a dependency.
```

- unmodified code: gives the idea I don't have to change my code to be compatible. This lead me to expect that I can pip install, add some plugin/bridge code and run.
- this is like six: gives me the idea I can choose in which reactor I run
